odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,546 @@
+        displacement : `domain` element
+            Point that the derivative need to be computed at.
+
+        Returns
+        -------
+        derivative : `LinDeforFixedTempDeriv`
+            The derivative evaluated at ``displacement``.
+        """
+        return LinDeformFixedTemplDeriv(self.template, displacement)
+
+
+class LinDeformFixedTemplDeriv(Operator):
+
+    """Derivative of the fixed template linearized deformation operator.
+
+    This operator computes the derivative of the fixed template
+    linearized deformation operator for the fixed
+    template ``I`` at given displacement ``v``:
+
+        u --> grad(I)(Id + v).T u
+
+    where the ``u`` is a given vector field and
+    the ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement field:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain.
+    """
+
+    def __init__(self, template, displacement):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLp` element
+            Fixed template deformed by the vector field.
+        displacement: `ProductSpace` element
+            Point that the derivative needs to be computed at.
+
+        Examples
+        --------
+        >>> import odl
+        >>> space = odl.uniform_discr(0, 1, 5)
+        >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+        >>> template = space.element([0, 0, 1, 0, 0])
+        >>> disp_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+        >>> vector_field = disp_field_space.element([[1, 1, 1, 2, 1]])
+        >>> op = LinDeformFixedTemplDeriv(template, disp_field)
+        >>> op(vector_field)
+        uniform_discr(0.0, 1.0, 5).element([0.0, 5.0, -5.0, -10.0, 0.0])
+        """
+        if displacement[0] not in template.space:
+            raise TypeError('`displacement[0]` {!r} not an element of'
+                            '`template.space`'.format(displacement[0],
+                                                      template.space))
+
+        Operator.__init__(self, displacement.space, template.space,
+                          linear=True)
+
+        self.template = template
+        self.displacement = displacement
+
+    def _call(self, vector_field):
+        """Implementation of ``self(vector_field)``.
+
+        Parameters
+        ----------
+        vector_field: `domain` element
+            The evaluation point, i.e. a displacement field, for the
+            deformed gradient of the template.
+        """
+        if vector_field not in self.displacement.space:
+            raise TypeError('`vector_field` {!r} not an element of'
+                            '`displacement.space`'.format(
+                                vector_field, self.displacement.space))
+
+        grad = Gradient(self.range, method='forward',
+                        padding_method='symmetric')
+        grad_template = grad(self.template)
```

Comment:
This is of course dependent on the use case. If we do iterative reconstruction, we usually need the forward once and the derivative adjoint once. So you're right, in that situation it will be sufficient if we can pass the `grad_template` to the adjoint, this will save us quite some computation.

On the other hand, if we need to evaluate the forward operator multiple times, caching will save time even there.

I don't exactly remember where the inconsistencies exactly came from. But if we pass a copy to the adjoint instead of a reference to the same array, there should be no problems.